### ISSUE TITLE
Potential issue when standard_name has no canonical units defined (eg. for soil_type, region, area_type)

### ISSUE THREAD

**Issue by sol1105** [ioos/compliance-checker#1219]
**Version of compliance checker running:** `5.3.0`

**Describe the checker this affects:** `cf:1.6`

**Attach a minimal CDL or NetCDF file which is able to reproduce the issue** https://swift.dkrz.de/v1/dkrz_485fa9f6d16841fc847fb61048428062/cf_test/testfile.nc?temp_url_sig=aae2df4f333142a18e82523b8995299aab772989&temp_url_expires=2025-10-02T19:33:20Z

**To Reproduce:**
`cchecker.py -t cf:1.6 testfile.nc`

**Describe the issue below:**
The standard_name `soil_type` has no canonical units defined, depending on what unit I set in the netCDF-file, cc states that:

`units=1`
>Errors
> §3.1 Units
> * Units "1" for variable SOILTYP must be convertible to canonical units "None"

`no units attribute`
>Errors
> §3.1 Units
> * units attribute is required for SOILTYP when variable is not a dimensionless quantity

The `soil_type` variable is defined as follows:
```
int SOILTYP(rlat, rlon) ;
       [...]
	SOILTYP:standard_name = "soil_type" ;
	SOILTYP:flag_values = 1, 2, 3, 4, 5, 6, 7, 8, 9, 0 ;
	SOILTYP:units = "1" ;
```

Should generally **units="1"** or a **missing units attribute** be accepted (= no error raised) if no canonical units are defined for a standard_name? Affected are  standard_names `region`, `soil_type`, `area_type` (aliases `surface_cover`, `land_cover`), `predominant_precipitation_type_at_surface`, `land_cover_lccs`, maybe more.

**Comment by benjwadams** [comment]
I haven't forgotten this issue, I just haven't found the relevant section of CF that deals with unitless variables.

**Comment by Shreyaav18** [comment]
I've analyzed this issue. The problem is that categorical variables like `soil_type`, `region`, and `area_type` have **no canonical units** defined in the CF standard name table (because they're classification codes, not physical measurements), but the checker doesn't handle this case.

**Proposed approach:**
- Detect when a standard name has no canonical units → treat as categorical variable
- For categorical variables: make `units` attribute optional
- Accept `units="1"` or omit units entirely
- Skip dimensional units validation for these variables

This would fix the contradictory errors while maintaining strict validation for normal dimensional variables.

Would this approach be acceptable? Happy to submit a PR if maintainers agree. @benjwadams

**Comment by AdityaPetkar2024** [comment]
Hi, I'd like to work on this fix. Based on @Shreyaav18's proposed approach, I plan to detect standard names with no canonical units and treat them as categorical variables, making units optional in that case. Would this approach still be acceptable? Happy to submit a PR.

**Comment by ocefpaf** [comment]
@AdityaPetkar2024 we need to find the relevant text in the CF first and, if not, open an issue upstream requesting clarification on the topic.

**Comment by AdityaPetkar2024** [comment]
Thanks for the guidance. I will look through the CF conventions and standard name table for relevant text on this. If the spec doesn't explicitly cover this case, should I open an issue at cf-convention/cf-conventions requesting clarification before this PR proceeds?

**Comment by ocefpaf** [comment]
> Thanks for the guidance. I will look through the CF conventions and standard name table for relevant text on this.

Yep. @Shreyaav18 already did some research. You can ask them for guidance in case you need help. Note that we will need to reference the paragraphs in the CF docs about this.

> If the spec doesn't explicitly cover this case, should I open an issue at cf-convention/cf-conventions requesting clarification before this PR proceeds?

Correct.

**Comment by AdityaPetkar2024** [comment]
@Shreyaav18 Hi, I'm working on the fix for this issue. ocefpaf mentioned you did some research on the CF spec — would you be able to share what you found? Specifically looking for any text in the CF conventions that covers how checkers should handle standard names with no canonical units defined.

**Comment by AdityaPetkar2024** [comment]
After reviewing the CF conventions, I found relevant justification in §3.3 and Example 3.7. The spec explicitly uses region and area_type as examples of standard names that indicate quantities taking only certain standard values (classification codes). Example 3.7 shows a variable with standard_name: region with no units attribute, only flag_values and flag_meanings. This confirms that categorical standard names are not expected to have dimensional units, justifying the fix. @ocefpaf does this citation satisfy the requirement, or should I still open an upstream issue at cf-convention/cf-conventions?

**Comment by ocefpaf** [comment]
> After reviewing the CF conventions, I found relevant justification in §3.3 and Example 3.7. The spec explicitly uses region and area_type as examples of standard names that indicate quantities taking only certain standard values (classification codes). Example 3.7 shows a variable with standard_name: region with no units attribute, only flag_values and flag_meanings. This confirms that categorical standard names are not expected to have dimensional units, justifying the fix. [@ocefpaf](https://github.com/ocefpaf) does this citation satisfy the requirement, or should I still open an upstream issue at cf-convention/cf-conventions?

Let's take this discussion to the PR. Also, please post the links to those paragraphs to make it easier for others to read and review. We may need to add comments to the code as well to remind us where these changes are from.
